mirgecom/diffusion.py

```diff
@@ -164,40 +176,122 @@ class NeumannDiffusionBoundary(DiffusionBoundary):
             interior=kappa_minus,
             exterior=kappa_minus)
         normal = actx.thaw(dcoll.normal(dd_bdry))
+        ext_value = self.value if self.function is None else \
+                    self.function(u_minus=u_minus, grad_u_minus=grad_u_minus,
+                                  **kwargs)
         grad_u_tpair = TracePair(dd_bdry,
             interior=grad_u_minus,
             exterior=(
                 grad_u_minus
-                + 2 * (self.value - np.dot(grad_u_minus, normal)) * normal))
+                + 2 * (ext_value - np.dot(grad_u_minus, normal)) * normal))
         return diffusion_facial_flux(kappa_tpair, grad_u_tpair, normal)
 
 
+class PrescribedFluxDiffusionBoundary(DiffusionBoundary):
+    r"""Prescribed flux boundary condition for the diffusion operator.
+
+    For the boundary condition $(\nabla u \cdot \mathbf{\hat{n}})|_\Gamma$,
+    uses external data
+
+    .. math::
+
+        u^+ = u^-
+
+    when computing the boundary fluxes for $\nabla u$, and and uses the
+    prescribed flux to evaluate the numerical flux $f^*$ using
+
+    .. math::
+
+        f^*|_\Gamma = \frac{1}{2} (F_{presc} + F^-)
+
+    when computing the boundary fluxes for $\nabla \cdot (\kappa \nabla u)$.
+
+    .. automethod:: __init__
+    .. automethod:: get_grad_flux
+    .. automethod:: get_diffusion_flux
+    """
+
+    def __init__(self, function):
+        """
+        Initialize the boundary condition.
+
+        Parameters
+        ----------
+        function
+            function prescribing the external flux
+        """
+        self.function = function
+
+    def get_grad_flux(self, dcoll, dd_bdry, u_minus, **kwargs):
+        """Flux for gradient evaluation."""
+        actx = u_minus.array_context
+        u_tpair = TracePair(dd_bdry,
+            interior=u_minus,
+            exterior=u_minus)
+        normal = actx.thaw(dcoll.normal(dd_bdry))
+        return grad_facial_flux(u_tpair, normal)
+
+    def get_diffusion_flux(self, dcoll, dd_bdry, u_minus, kappa_minus,
+                           grad_u_minus, **kwargs):
+        """Flux for Laplacian evaluation."""
+        actx = grad_u_minus[0].array_context
+        u_tpair = TracePair(dd_bdry,
+            interior=u_minus,
+            exterior=u_minus)
+        kappa_tpair = TracePair(dd_bdry,
+            interior=kappa_minus,
+            exterior=kappa_minus)
+        grad_u_tpair = TracePair(dd_bdry,
+            interior=grad_u_minus,
+            exterior=grad_u_minus)
+        normal = actx.thaw(dcoll.normal(dd_bdry))
+
+        external_flux = self.function(u_tpair=u_tpair,
+            kappa_tpair=kappa_tpair, grad_u_tpair=grad_u_tpair,
+            normal=normal, **kwargs)
+
+        # average between the prescribed value and the internal value
+        # FIXME maybe the minus in the prescribed function is due to the normal?
+        return 0.5*(-kappa_tpair.int * np.dot(grad_u_tpair.int, normal)
+            - external_flux)
```

Comment:
My guess is the minus comes from the fact that the diffusion operator uses the physical flux `-kappa*grad_u` as the form of the numerical flux, not `kappa*grad_u` like the book. Can't say for sure without seeing the code that constructs `function`.